src/formats/MOL2.cpp

```diff
@@ -0,0 +1,298 @@
+
+    long long natoms = string2longlong(counts[0]);
+    long long nbonds = 0;
+    
+    if (counts.size() >= 2) {
+        nbonds = string2longlong(counts[1]);
+    }
+
+    residues_.clear();
+    frame.resize(0);
+    frame.reserve(static_cast<size_t>(natoms));
+
+    auto junk = file_->readline();
+    bool charges = (trim(file_->readline()) != "NO_CHARGES");
+
+    while (!file_->eof()) {
+        auto curr_pos = file_->tellg();
+        auto line2 = trim(file_->readline());
+
+        if (line2 == "@<TRIPOS>ATOM") {
+            read_atoms(frame, static_cast<size_t>(natoms), charges);
+        } else if (line2 == "@<TRIPOS>BOND") {
+            read_bonds(frame, static_cast<size_t>(nbonds));
+        } else if (line2 == "@<TRIPOS>CRYSIN") {
+            auto cryst = file_->readline();
+
+            double a, b, c, alpha, beta, gamma;
+            scan(cryst, " %lf %lf %lf %lf %lf %lf",
+                &a, &b, &c, &alpha, &beta, &gamma
+            );
+
+            frame.set_cell(UnitCell(a, b, c, alpha, beta, gamma));
+        } else if (line2 == "@<TRIPOS>MOLECULE") {
+            file_->seekg(curr_pos);
+            break;
+        }
+    }
+
+    for (auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+
+}
+
+void MOL2Format::read_atoms(Frame& frame, size_t natoms, bool charges) {
+    auto lines = file_->readlines(natoms);
+
+    for (const auto& line : lines) {
+        unsigned long id, resid;
+        char atom_name[32], atom_type[32], res_name[32];
+        double x, y, z;
+        double charge = 0;
+
+        if (charges) {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s %lf",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0], &charge
+            );
+        } else {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0]
+            );
+        }
+
+        Atom atom(atom_name);
+        frame.add_atom(std::move(atom), Vector3D(x, y, z));
+        size_t current_atom = frame.size() - 1;
+        if (charges) {
+            frame[current_atom].set_charge(charge);
+        }
+        frame[current_atom].set_type(atom_type);        
+
+        if (residues_.find(resid) == residues_.end()) {
+            Residue residue(res_name, resid);
+            residue.add_atom(current_atom);
+
+            residues_.insert({resid, residue});
+        } else {
+            // Just add this atom to the residue
+            residues_.at(resid).add_atom(current_atom);
+        }
+
+    }
+}
+
+void MOL2Format::read_bonds(Frame& frame, size_t nbonds) {
+    auto lines = file_->readlines(nbonds);
+
+    for (const auto& line : lines) {
+        unsigned long id, id_1, id_2;
+
+        scan(line, " %lu %lu %lu", &id, &id_1, &id_2);
+
+        // MOL2 is 1 index-based, not 0
+        --id_1;
+        --id_2;
+
+        if (id_1 >= frame.size() || id_2 >= frame.size()) {
```

Comment:
Does the format specify that the bonds section appears after the atoms section? Else this test would fail